### PR TITLE
Refactor ad blocking recovery tag storage to use `Setting` class.

### PR DESCRIPTION
**Pull request by kuasha420**
## Summary

Addresses issue:

- #7156

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Modules/AdSense/Ad_Blocking_Recovery_Tag.php
Comment (eugene-manuilov) [inline includes/Modules/AdSense/Ad_Blocking_Recovery_Tag.php:36]: Basically what we do for Ad Blocking Recovery can be easily considered as XSS since we inject a 3rd party script on frontend pages. We must be extremely careful here to make sure that what we get from the option is not a malicious code that a hacker managed to save into the database.

Base64 encoding is a very bad choice for encryption here because anyone can use it to create a malicious script and save it into the database that we will inject on each page. We must use the `Data_Encryption` class to really encrypt these values. Probably the best option for this class would be to override constructor to wrap passed `$options` with a new instance of the `Encrypted_Options` class and then pass the new encrypted options to the parent constructor.

cc @aaemnnosttv

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @kuasha420. Left a few comments for you. We need to polish this PR a bit before we can merge it. Please, take a look.

File: includes/Modules/AdSense.php
Comment (aaemnnosttv) [reply includes/Modules/AdSense.php]: This method is called for in the AC but since it's only needed by its tag class(es) we shouldn't need it yet. I'll update the AC accordingly.

File: includes/Modules/AdSense.php
Comment (aaemnnosttv) [inline includes/Modules/AdSense.php:84]:
```suggestion
	 * 
	 * @var Ad_Blocking_Recovery_Tag
```

File: includes/Modules/AdSense/Ad_Blocking_Recovery_Tag.php
Comment (tofumatt) [reply includes/Modules/AdSense/Ad_Blocking_Recovery_Tag.php:37]: My perspective is that in order for someone/something to inject code into the string we output, they already need to have access to the database somehow, because the tag is fetched and saved on the server, not from user input. So there needs to be:

* a man-in-the-middle between Site Kit’s server and Google, OR
* database access

for an attacker to write to this string.

At that point, there are LOTS of other vectors for users to inject `<script>` tags, etc. They can place it in `post content`, etc. I suppose the only thing we’re doing here is making sure that our echo statement isn’t the one outputting bad content, but if a user has enough access to read encryption keys and database records, etc. they can manage to replace the data we store.

Ultimately someone has to have so much access to mess with that data that it’s game over anyway, so it feels like we’re trying to prevent anything bad from happening in Site Kit when an attacker can already do way worse things. So the added complexity of encryption, hashing, etc. for data that needs to be read on every page render and needs to be decrypted seemed a bit misguided and complex for very limited benefit.

The base64 handling wasn’t to encrypt so much as just prevent any automated scripts/plugins that went through database records from messing with anything that looks like a `<script>` tag. I could even see misguided “security” plugins searching for these strings inside DB records and thinking they ARE hacking attempts, and removing them, so the base64 approach was just about obscuring things enough to prevent silly overwrites.

But ultimately, the nature of this tag—because it’s always-changing from AdSense and thus essentially needs to be output as a raw, unverifiable/checkable string, means that we have limited options.

Hashing it and verifying that hash before rendering (or using the encryption library) would prevent someone from messing with that field to output malicious HTML, but there are other places that are better spots to output that HTML (eg. `post_content`).

I don’t mean to hand-wave away security, but in order for this to even be an attractive attack vector the site would need:

1. Site Kit
2. AdSense
3. Ad Blocking Recovery
4. A plugin/database vulnerability that allows arbitrary write access to database fields (which functionality is already game-over for their site)
5. A security plugin that restricts post_content from outputting <script> tags thus making that an unviable attack vector but that also doesn’t restrict other HTML output from plugins

I think by the time we’re in trouble here, it’s already game-over for that site. If another plugin wants to write arbitrary SQL to a website it can.

It doesn't _really_ matter if a bad actor writes a BitCoin miner `<script>` tag into the HTML _we_ output or `post_content` outputs. I suppose the main concern would be a vulnerability in our plugin that allows users to write specifically to that option from the client-side. If that's the concern, then I suppose we do need to encrypt it. But otherwise I think the end result is inescapable when there is the ability for _any other plugin_ to write _any content to the database_. 🤷🏻‍♂️

Comment (eugene-manuilov) [reply includes/Modules/AdSense/Ad_Blocking_Recovery_Tag.php:37]: > I suppose the main concern would be a vulnerability in our plugin that allows users to write specifically to that option from the client-side. If that's the concern, then I suppose we do need to encrypt it.

@tofumatt @aaemnnosttv, yes, but not in our plugin. That vulnerability can be in any 3rd party plugin or theme that would let a bad guy save options with an arbitrary name from the client side (for example, poorly secured custom REST API endpoint). In this case, they can use that plugin/theme to save a malicious script with our option name and it will be rendered by our plugin on the frontend -> we will be guilty for XSS injections. If we use `Encrypted_Options`, then they won't be able to use this approach to fake our script.

> Also, if we do decide to use encryption, this would not require changes to `Ad_Blocking_Recovery_Tag` since it accepts an `Options_Interface` as its constructor dependency, so we could encrypt the entire value just by passing an instance of `Encrypted_Options` instead.

Yes, this is kind of the same what I suggested. It can be done either way: in the class that instantiates an instance of the `Ad_Blocking_Recovery_Tag` class or in the constructor of the class itself if we still want to pass just `$options` from the AdSense class (that doesn't have an instance of encrypted options and needs to create one too).

Comment (tofumatt) [reply includes/Modules/AdSense/Ad_Blocking_Recovery_Tag.php:37]: Fair enough! I suppose it's relatively harmless for us to encrypt that data to prevent that attack vector. Thanks for catching this one @eugene-manuilov 👍🏻

File: includes/Modules/AdSense.php
Comment (kuasha420) [reply includes/Modules/AdSense.php:84]: This is consistent with how it's done on the base `Module` class.

Comment (aaemnnosttv) [reply includes/Modules/AdSense.php:84]: Fair enough; feel free to leave as-is. We've never been entirely consistent about it. There are explicit rules that just haven't made it into PHPCS yet. See #924

> * For methods and functions, tags should be grouped: `since`/`deprecated`/`access`/`static` should be grouped together, separated by a blank line from `param`/`return` which should be grouped together. If there is `global`, it should be separated from both others by a blank line (essentially be its own group).

Regarding `var` it's order is defined as coming after `global` so it seems that it would be implicitly separated from the above grouping with `@since`.

Again, just for context; don't worry about it 😉

**Review by eugene-manuilov — APPROVED** [review]

**Pull request merged**
